vowpalwabbit/interactions_predict.h

```diff
@@ -94,6 +134,184 @@ inline void inner_kernel(DataT& dat, features::const_audit_iterator& begin, feat
+{
+  size_t num_features = 0;
+  auto first_begin = std::get<0>(range).first;
+  const auto& first_end = std::get<0>(range).second;
+  const auto& second_begin = std::get<1>(range).first;
+  auto& second_end = std::get<1>(range).second;
+
+  const bool same_namespace = (!permutations && (first_begin == second_begin));
+  size_t i = 0;
+  for (; first_begin != first_end; ++first_begin)
+  {
+    feature_index halfhash = FNV_prime * first_begin.index();
+    if (Audit) { audit_func(first_begin.audit() != nullptr ? first_begin.audit() : &EMPTY_AUDIT_STRINGS); }
+    // next index differs for permutations and simple combinations
+    auto begin = second_begin;
+    if (same_namespace) { begin += i; }
+    num_features += std::distance(begin, second_end);
+    kernel_func(begin, second_end, first_begin.value(), halfhash);
+    if (Audit) audit_func(nullptr);
+    i++;
+  }
+  return num_features;
+}
+
+template <bool Audit, typename KernelFuncT, typename AuditFuncT>
+size_t process_cubic_interaction(
+    const std::tuple<std::pair<features::const_audit_iterator, features::const_audit_iterator>,
+        std::pair<features::const_audit_iterator, features::const_audit_iterator>,
+        std::pair<features::const_audit_iterator, features::const_audit_iterator>>& range,
+    bool permutations, const KernelFuncT& kernel_func, const AuditFuncT& audit_func)
+{
+  size_t num_features = 0;
+  auto first_begin = std::get<0>(range).first;
+  const auto& first_end = std::get<0>(range).second;
+  const auto& second_begin = std::get<1>(range).first;
+  auto second_end = std::get<1>(range).second;
+  const auto& third_begin = std::get<2>(range).first;
+  auto& third_end = std::get<2>(range).second;
+
+  // don't compare 1 and 3 as interaction is sorted
+  const bool same_namespace1 = (!permutations && (first_begin == second_begin));
+  const bool same_namespace2 = (!permutations && (second_begin == third_begin));
+
+  size_t i = 0;
+  for (; first_begin != first_end; ++first_begin)
+  {
+    if (Audit) { audit_func(first_begin.audit() != nullptr ? first_begin.audit() : &EMPTY_AUDIT_STRINGS); }
+
+    const uint64_t halfhash1 = FNV_prime * first_begin.index();
+    const float first_ft_value = first_begin.value();
+    size_t j = 0;
+    if (same_namespace1)  // next index differs for permutations and simple combinations
+    { j = i; }
+
+    for (auto inner_second_begin = second_begin + j; inner_second_begin != second_end; ++inner_second_begin)
+    {
+      // f3 x k*(f2 x k*f1)
+      if (Audit)
+      { audit_func(inner_second_begin.audit() != nullptr ? inner_second_begin.audit() : &EMPTY_AUDIT_STRINGS); }
+      feature_index halfhash = FNV_prime * (halfhash1 ^ inner_second_begin.index());
+      feature_value ft_value = INTERACTION_VALUE(first_ft_value, inner_second_begin.value());
+
+      auto begin = third_begin;
+      // next index differs for permutations and simple combinations
+      if (same_namespace2) { begin += j; }
```

Comment:
is it possible for `first_begin == second_begin == third_begin`? In this case, would it be possible to iterate begin past `third_end`?

Reply:
I'm not changing the logic but they would be equal on the first iteration, and then never again after that